stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/io/channels/base/Channel.java

```diff
@@ -98,51 +206,37 @@ public Channel(ByteChannel channel, int size) throws BallerinaIOException {
      *
      * @param numberOfBytes the number of bytes required.
      * @return bytes which are retrieved from the channel.
-     * @throws BallerinaIOException during I/O error.
+     * @throws IOException during I/O error.
      */
-    public byte[] read(int numberOfBytes) throws BallerinaIOException {
+    public byte[] readFull(int numberOfBytes) throws IOException {
         ByteBuffer readBuffer = contentBuffer.get(numberOfBytes, this);
-        return readBuffer.array();
+        byte[] content = readBuffer.array();
+        int contentLength = readBuffer.capacity();
+        if (content.length > numberOfBytes || contentLength < numberOfBytes) {
+            //We need to collect the bytes only belonging to the offset, since the number of bytes returned can be < the
+            //required amount of bytes
+            content = Arrays.copyOfRange(content, 0, contentLength);
+        }
+        return content;
     }
 
     /**
-     * Reads all bytes from the I/O source.
+     * Closes the given channel.
      *
-     * @return all the bytes read.
-     * @throws BallerinaException during I/O error.
+     * @throws IOException errors occur while closing the connection.
      */
-    public byte[] readAll() throws BallerinaException {
-        List<ByteBuffer> readBufferList = new ArrayList<>();
-        int totalNumberOfBytes = 0;
-        boolean hasRemaining = false;
-        do {
-            ByteBuffer readBuffer = contentBuffer.get(MAX_BUFFER_CHUNK_SIZE, this);
-            int numberOfBytesRead = readBuffer.limit();
-            if (numberOfBytesRead > 0) {
-                readBufferList.add(readBuffer);
-                totalNumberOfBytes = totalNumberOfBytes + numberOfBytesRead;
+    public void close() throws IOException {
+        try {
+            if (null != channel) {
+                channel.close();
             } else {
-                hasRemaining = true;
+                log.error("The channel has already being closed");
             }
-        } while (!hasRemaining);
-        return merge(readBufferList, totalNumberOfBytes);
-    }
-
-    /**
-     * Writes bytes to channel.
-     *
-     * @param content     the data which will be written.
-     * @param startOffset if the data should be written from an offset (starting byte position).
-     * @return the number of bytes written.
-     */
-    public int write(byte[] content, int startOffset) throws BallerinaException {
-        ByteBuffer outputBuffer = ByteBuffer.wrap(content);
-        //If a larger position is set, the position would be disregarded
-        outputBuffer.position(startOffset);
-        if (log.isDebugEnabled()) {
-            log.debug("Writing " + content.length + " for the buffer with offset " + startOffset);
+        } catch (IOException e) {
+            String message = "Error occurred while closing the connection";
+            log.error(message, e);
```

Comment:
Single line?

Reply:
This way is more readable and extensible IMO. Let's say we want to log the message. If so instead of duplicating we could use the message itself